### ISSUE TITLE
Feature request - allow reuse of Gradients

### ISSUE THREAD

**Issue by fuzing** [tracel-ai/burn#5687]
It would be useful to be able to reuse Gradients - For example (plain step function):
```
    fn step(&self, item: FlowRegFormerTrainingBatch) -> TrainOutput<RegressionOutput> {
        // Run forward pass, calculate gradients and return them along with the output
        let item = self.forward(item);
        let grads = item.loss.backward();
        TrainOutput::new(self, grads, item)
    }
```

I'm attempting to instrument/log gradient magnitudes, and find something like this useful:
```
    fn step(&self, item: FlowRegFormerTrainingBatch) -> TrainOutput<RegressionOutput> {
        // Run forward pass, calculate gradients and return them along with the output
        let item = self.forward(item);
        let grads = item.loss.backward();
        
        // gradient examination
        let model = self;
        let grads_params = GradientsParams::from_grads(grads, model);
        let mut inspector = GradientInspector::new(&grads_params, true);
        model.visit(&mut inspector);
        panic!("completed");
        // end gradient examination

        TrainOutput::new(self, grads, item)
    }
```

The problem is that ``GradientsParams::from_grads()`` consumes the gradients, which subsequently need to be returned from the step method in the TrainOutput struct.

My current workaround is to panic! after examining the gradients,  meaning I can only examine them once at the start of an epoch (i.e. stop training then restart from last epoch -> examine gradients -> panic!).

It would be useful if the Gradients were clonable/copyable/reusable, or if GradientParams::from_grads() took a reference to the generated gradients and didn't consume them.

Thank you

**Comment by laggui** [comment]
You can inspect the gradients after constructing `TrainOutput`, which exposes the converted `GradientsParams` through its public grads field:

```rust
fn step(&self, item: FlowRegFormerTrainingBatch) -> TrainOutput<RegressionOutput> {
    let item = self.forward(item);
    let grads = item.loss.backward();
    let output = TrainOutput::new(self, grads, item);

    let mut inspector = GradientInspector::new(&output.grads, true);
    self.visit(&mut inspector);

    output
}
```

As long as your inspector reads gradients using `GradientsParams::get()` rather than removing them, they remain available for the optimizer update. No cloning or interruption of training is needed.

You can also keep your explicit conversion, but you'd have to return the result directly:
```rust
let grads_params = GradientsParams::from_grads(grads, self);
self.visit(&mut GradientInspector::new(&grads_params));

TrainOutput {
    grads: grads_params,
    item,
}
```

**Comment by fuzing** [comment]
Of course - got it - thank you @laggui - much appreciated!

**Comment by laggui** [comment]
No problem!

I'll close the issue since the existing API already supports the use case.
